<a href="https://colab.research.google.com/github/PeedroModa/An-lise_Diabetes/blob/main/An%C3%A1lise_Diabetes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Problema a ser resolvido**

###### O time de negócios está interessado em descobrir padrões e insights que os ajudem a entender melhor o controle do diabetes e os fatores de risco associados. Eles querem identificar relações entre os diferentes atributos do dataset e a ocorrência de diabetes, a fim de desenvolver estratégias de prevenção mais eficazes, melhorar o tratamento dos pacientes e otimizar os recursos médicos disponíveis.

Descrição do dataset:

- Pregnancies: Número de gestações.
- Glucose: Nível de glicose no sangue.
- BloodPressure: Medição da pressão arterial.
- SkinThickness: Espessura da pele.
- Insulin: Nível de insulina no sangue.
- BMI: Índice de massa corporal.
- DiabetesPedigreeFunction: Probabilidade de diabetes com base no histórico familiar.
- Age: Idade
- Outcome: Diabético ou não (1 para sim e 0 para não)

# Importação de bibliotecas

In [ ]:
import pandas as pd
import sqlite3


In [ ]:
# Carregar arquivo CSV

caminho = "<input|output_path_here>"
dados = pd.read_csv(caminho)

In [ ]:
# Criar conexão e conectar com um banco SQLite em memória

conn = sqlite3.connect(':memory:')

In [ ]:
# Carregar os dados do df para uma tabela em SQL

dados.to_sql('diabetes', conn, index=False)

2460

In [ ]:
# Consulta para visualização das primeiras 5 linhas do dataset

consulta_db = "SELECT * FROM diabetes LIMIT 5"
resultado_consulta_db = pd.read_sql_query(consulta_db, conn)
resultado_consulta_db

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

# EDA (Análise exploratória de dados).

###### OBS: .describe e .info fiz utilizando python pois fica melhor o entendimento dos dados e visualização geral das informações.

In [ ]:
# Informações do dataset

dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2460 entries, 0 to 2459
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               2460 non-null   int64  
 1   Glucose                   2460 non-null   int64  
 2   BloodPressure             2460 non-null   int64  
 3   SkinThickness             2460 non-null   int64  
 4   Insulin                   2460 non-null   int64  
 5   BMI                       2460 non-null   float64
 6   DiabetesPedigreeFunction  2460 non-null   float64
 7   Age                       2460 non-null   int64  
 8   Outcome                   2460 non-null   int64  
dtypes: float64(2), int64(7)
memory usage: 173.1 KB


#### Informações:

- O dataset possui 2460 valores de entrada.
- Possui 9 colunas em todo dataframe.
- Em todas as colunas não existem valores nulos.
- Há dois tipos de dados, int64 e float64, sendo int64 valores intereiros e float54 valores decimais, indicando porcentagem.


In [ ]:
## Estatística descritiva do dataframe:

dados.describe().T

count        mean         std     min        25%  \
Pregnancies               2460.0    3.817480    3.296458   0.000    1.00000   
Glucose                   2460.0  121.602033   31.789270   0.000  100.00000   
BloodPressure             2460.0   68.915041   19.082655   0.000   64.00000   
SkinThickness             2460.0   20.531301   15.716901   0.000    0.00000   
Insulin                   2460.0   80.119919  116.765807   0.000    0.00000   
BMI                       2460.0   31.990447    7.802569   0.000   27.10000   
DiabetesPedigreeFunction  2460.0    0.491440    0.363917   0.078    0.25175   
Age                       2460.0   32.821951   11.251208  21.000   24.00000   
Outcome                   2460.0    0.390244    0.487904   0.000    0.00000   

                              50%      75%     max  
Pregnancies                 3.000    6.000   17.00  
Glucose                   117.000  142.000  197.00  
BloodPressure              70.000   80.000  122.00  
SkinThickness              23.000   33.000   63.00  
Insulin                    36.000  129.000  846.00  
BMI                        32.100   36.500   67.10  
DiabetesPedigreeFunction    0.381    0.647    2.42  
Age                        29.000   39.000   81.00  
Outcome                     0.000    1.000    1.00

#### Informações:

- Pregnancies: O número médio de gravidezes é 3.8, variando de 0 a 17.
- Glucose: O nível médio de glicose no sangue é aproximadamente 121.6 mg/dL. Se a aferição ocorreu em jejum, pode-se considerar que grande parte dos usuários são pré-diabéticos.
- BloodPressure: A média de pressão do sangue é de aproximadamente 68.9 mmHg. Considerando que esse valor se refere a pressão arterial diastólica (PAD), podemos considerar que a média está dentro dos valores normais para os usuários saudáveis acima dos 18 anos. A amostra nos mostra um desvio padrão de 19 com valores que vão de 0 a 122mmHg.
- SkinThickness: A média de espessura da pele é de 20.5mm, com um desvio padrão de 15mm, variando de 0 a 63mm.
- Insulin: A média de insulina é de 80.12 ui/ml (0,8 ml), tendo como valor minimo 0 e valor máximo 846 ui/ml.
- BMI: A média do índice de massa corporal é de aproximadamente 31.99, sendo considerado obesidade de grau 1. Seu valores variam de 0 a 67.10.
- DiabetesPedigreeFunction: A porcentagem média de aquisição de diabetes devido a histórico familiar é de 0.49, tendo como valor minimo 0.07 e valor máximo 2.42.
- Age: A idade média do grupo de usuários de é de 32 anos, possuindo um desvio padrão de 11.25. A pessoa com a menor idade entre o grupo possui 21 anos e a mais velha possui 81 anos.
- Outcome: Essa variável indica que aproximadamente 39% dos usuários deste dataset possuem diabetes.


#### Dados inconsistentes:

- De acordo com as informações vistas anteriormente, pode-se notar uma inconsistencia nos dados, pois algumas colunas possuiam valores mínimos = 0, o que demonstra ser um dado errôneo.

- As seguintes colunas possuem valores mínimos iguais a zero, ou seja, valores irreais para um ser humano: BloodPressure, SkinThickness, Insulin, BMI.


# Análise univariada.

In [ ]:
# Quais as 10 idades que mais aparecem no dataset?

d1 = ''' SELECT Age, COUNT(*) as QtAge
          FROM diabetes
          GROUP BY Age
          ORDER BY QtAge DESC
          LIMIT 10 '''

re_d1 = pd.read_sql_query(d1, conn)
re_d1

Age  QtAge
0   21    214
1   22    203
2   25    183
3   24    135
4   29    122
5   23    113
6   26    101
7   27     98
8   28     93
9   41     85

In [ ]:
# Qual a distribuição de gravidez entre os indivíduos?

d2 = ''' SELECT Pregnancies, COUNT(*) AS QtPreg, ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM diabetes)) AS Percentual
          FROM diabetes
          GROUP BY Pregnancies
          ORDER BY QtPreg DESC
          LIMIT 10'''

re_d2 = pd.read_sql_query(d2, conn)
re_d2

Pregnancies  QtPreg  Percentual
0            1     423        17.0
1            0     350        14.0
2            2     326        13.0
3            5     261        11.0
4            4     245        10.0
5            3     212         9.0
6            7     170         7.0
7            6     117         5.0
8            8     114         5.0
9            9      87         4.0

In [ ]:
# De acordo com os valores de Glicemia, qual a quantidade de pessoas com hipoglicemia, saudáveis, pré-diabéticos e diabéticos?

d3 = ''' SELECT sub1.Classificação, COUNT(*) AS QtClassif, ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM diabetes)) AS Percentual
          FROM (
            SELECT Glucose,
          CASE
            WHEN Glucose >= 0 AND Glucose <= 69 THEN 'Hipoglicêmico'
            WHEN Glucose >= 70 AND Glucose <= 99 THEN 'Saudável'
            WHEN Glucose >= 100 AND Glucose <= 125 THEN 'Pré-diabético'
            ELSE 'Diabético'
        END AS Classificação
        FROM diabetes
        )as sub1
      GROUP BY sub1.Classificação
      ORDER BY QtClassif DESC'''

re_d3 = pd.read_sql_query(d3, conn)
re_d3

Classificação  QtClassif  Percentual
0      Diabético        966        39.0
1  Pré-diabético        931        38.0
2       Saudável        522        21.0
3  Hipoglicêmico         41         2.0

In [ ]:
# De acordo com os valores de massa corporal (BMI), quantas pessoas são abaixo do peso, peso normal, acima do peso, obesidade 1, obesidade 2, obesidade 3?

d4 = ''' SELECT sub1.Classificação, COUNT(*) AS QtClassif, ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM diabetes)) AS Percentual
          FROM (
            SELECT BMI,
          CASE
            WHEN BMI >= 0 AND BMI <= 18.4 THEN 'Abaixo do peso'
            WHEN BMI >= 18.5 AND BMI <= 24.8 THEN 'Peso normal'
            WHEN BMI >= 24.9 AND BMI <= 29.8 THEN 'Acima do peso'
            WHEN BMI >= 29.9 AND BMI <= 34.8 THEN 'Obesidade grau 1'
            WHEN BMI >= 34.9 AND BMI <= 39.8 THEN 'Obesidade grau 2'
            ELSE 'Obesidade grau 3'
        END AS Classificação
        FROM diabetes
        )as sub1
      GROUP BY sub1.Classificação
      ORDER BY QtClassif DESC'''

re_d4 = pd.read_sql_query(d4, conn)
re_d4

Classificação  QtClassif  Percentual
0  Obesidade grau 1        745        30.0
1     Acima do peso        516        21.0
2  Obesidade grau 2        485        20.0
3       Peso normal        336        14.0
4  Obesidade grau 3        327        13.0
5    Abaixo do peso         51         2.0

In [ ]:
# Quantas pessoas são diabéticas? Quantas não são?

d5 = ''' SELECT Outcome,
          COUNT(*) AS QtOutcome,
          ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM diabetes)) AS Percentual
          FROM diabetes
          GROUP BY Outcome
      '''

re_d5 = pd.read_sql_query(d5, conn)
re_d5

Outcome  QtOutcome  Percentual
0        0       1500        61.0
1        1        960        39.0

# Conclusões a partir das consultas anteriores:

- Age: A partir da consulta, é possível observar que 51% dos usuários do dataset estão dentro da faixa etária dos 21 ~ 29 anos.

- Pregnancies: A maioria dos usuários possui apenas 1 filho.

- Glucose: Observa-se que grande parte dos indivíduos do dataset são diabéticos ou pré-diabéticos, sendo aproximadamente 71%.

- BMI: A maioria dos usuários do dataset possui peso acima da taxa recomendada. Aproximadamente 71% dos usuários estão acima do peso.

- Segundo os dados do dataset, 61% dos usuários não possuem diabetes e 39% possuem.

- De acordo com os resultados das consultas anteriores e a alta prevalência de diabéticos e pré-diabéticos na amostra, podemos considerar o IMC como uma variável importante na obtenção de diabétes. Há de se observar que a maioria dos usuários estão acima do peso, com alguns apresentando obesidade de grau 1 e grau 2, sugerindo que o IMC elevado tem forte relação com um maior risco de desenvolver diabetes.




# Análise bivariada.

In [ ]:
# Quais são as 10 idades que mais registraram casos para diabetes?

f1 = ''' SELECT Age, COUNT(*) As QtUsuDiabetes
          FROM diabetes
          WHERE Outcome = 1
          GROUP BY Age
          ORDER BY QtUsuDiabetes DESC
          LIMIT 10 '''

re_f1 = pd.read_sql_query(f1, conn)
re_f1

Age  QtUsuDiabetes
0   25             75
1   29             62
2   41             59
3   31             53
4   33             45
5   38             41
6   36             41
7   23             39
8   28             38
9   43             37

In [ ]:
# Quais são as 10 idades que menos registraram casos para diabetes?

f2 = ''' SELECT Age, COUNT(*) As QtUsuDiabetes
          FROM diabetes
          WHERE Outcome = 1
          GROUP BY Age
          ORDER BY QtUsuDiabetes ASC
          LIMIT 10 '''

re_f2 = pd.read_sql_query(f2, conn)
re_f2

Age  QtUsuDiabetes
0   67              2
1   34              3
2   57              3
3   53              4
4   61              4
5   59              5
6   66              5
7   50              7
8   45              8
9   56              8

In [ ]:
# O número de gestações, influencia na obtenção de diabetes?


f3 = '''
      SELECT
    Pregnancies,
    COUNT(*) AS QtPreg,
    SUM(CASE WHEN Outcome = 1 THEN 1 ELSE 0 END) AS QtDiabetes,

    ROUND(SUM(CASE WHEN Outcome = 1 THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 2) AS Percentual

FROM
    diabetes
GROUP BY
    Pregnancies
ORDER BY Percentual DESC
           '''

re_f3 = pd.read_sql_query(f3, conn)
re_f3




Pregnancies  QtPreg  QtDiabetes  Percentual
0            17       6           6      100.00
1            15       5           5      100.00
2            14      10          10      100.00
3            11      19          16       84.21
4             8     114          71       62.28
5            12      24          14       58.33
6             9      87          50       57.47
7            10      63          34       53.97
8            13      28          15       53.57
9             7     170          91       53.53
10            5     261         117       44.83
11            0     350         143       40.86
12            4     245          95       38.78
13            3     212          80       37.74
14            6     117          36       30.77
15            2     326          79       24.23
16            1     423          98       23.17

In [ ]:
# Existe uma correlação entre os níveis de glicose no sangue e a pressão arterial em pacientes diabéticos?

f4 = ''' SELECT
    (COUNT(*) * SUM(Glucose * BloodPressure) - SUM(Glucose) * SUM(BloodPressure)) /
    (SQRT((COUNT(*) * SUM(Glucose * Glucose) - SUM(Glucose) * SUM(Glucose))) *
     SQRT((COUNT(*) * SUM(BloodPressure * BloodPressure) - SUM(BloodPressure) * SUM(BloodPressure)))) AS Correlation
FROM
    diabetes
WHERE
    Outcome = 1
      '''

re_f4 = pd.read_sql_query(f4, conn)
re_f4

Correlation
0     0.032099

In [ ]:
# Como a espessura da pele (SkinThickness) está relacionada com o índice de massa corporal (IMC) em pacientes diabéticos?

f5 = '''
        SELECT
        (COUNT(*) * SUM(SkinThickness * BMI) - SUM(SkinThickness) * SUM(BMI)) /
        (SQRT((COUNT(*) * SUM(SkinThickness * SkinThickness) - SUM(SkinThickness) * SUM(SkinThickness))) *
        SQRT((COUNT(*) * SUM(BMI * BMI) - SUM(BMI) * SUM(BMI)))) AS Correlation
        FROM diabetes
        WHERE Outcome = 1
        '''

re_f5 = pd.read_sql_query(f5, conn)
re_f5

Correlation
0      0.30397

In [ ]:
# Quais são os níveis médios de insulina em pacientes diabéticos e como eles variam de acordo com o IMC?


f6 = ''' SELECT sub1.Classificação,
        AVG(sub1.Insulin) AS Média_Insulina
          FROM(
            SELECT BMI, Insulin,
          CASE
            WHEN BMI >= 0 AND BMI <= 18.4 THEN 'Abaixo do peso'
            WHEN BMI >= 18.5 AND BMI <= 24.8 THEN 'Peso normal'
            WHEN BMI >= 24.9 AND BMI <= 29.8 THEN 'Acima do peso'
            WHEN BMI >= 29.9 AND BMI <= 34.8 THEN 'Obesidade grau 1'
            WHEN BMI >= 34.9 AND BMI <= 39.8 THEN 'Obesidade grau 2'
            ELSE 'Obesidade grau 3'
        END AS Classificação
        FROM diabetes
        WHERE Outcome = 1
        ) as sub1
        GROUP BY sub1.Classificação
      '''

re_f6 = pd.read_sql_query(f6, conn)
re_f6

Classificação  Média_Insulina
0    Abaixo do peso        0.000000
1     Acima do peso       68.263889
2  Obesidade grau 1      100.362434
3  Obesidade grau 2      119.474419
4  Obesidade grau 3       94.340206
5       Peso normal       19.636364

In [ ]:
# Existe uma relação entre a idade dos pacientes e a espessura da pele (SkinThickness) em pacientes diabéticos?

f7 = '''

SELECT
        (COUNT(*) * SUM(SkinThickness * Age) - SUM(SkinThickness) * SUM(Age)) /
        (SQRT((COUNT(*) * SUM(SkinThickness * SkinThickness) - SUM(SkinThickness) * SUM(SkinThickness))) *
        SQRT((COUNT(*) * SUM(Age * Age) - SUM(Age) * SUM(Age)))) AS Correlation
        FROM diabetes
        WHERE Outcome = 1

'''

re_f7 = pd.read_sql_query(f7, conn)

re_f7

Correlation
0     -0.16333

In [ ]:
# Quais são os padrões de espessura da pele em pacientes diabéticos e não diabéticos?

f8 = '''

SELECT
    CASE
        WHEN Outcome = 1 THEN 'Diabético'
        ELSE 'Não Diabético'
    END AS Categoria,
    AVG(SkinThickness) AS Média,
    MIN(SkinThickness) AS Mínimo,
    MAX(SkinThickness) AS Máximo
FROM
    diabetes
WHERE
    Outcome IN (0, 1)
GROUP BY
    Outcome
'''

re_f8 = pd.read_sql_query(f8, conn)

re_f8

Categoria      Média  Mínimo  Máximo
0  Não Diabético  19.637333       0      60
1      Diabético  21.928125       0      63

In [ ]:
# Existe uma correlação entre os níveis de insulina no sangue e a idade dos pacientes diabéticos?

f9 = '''

SELECT
        (COUNT(*) * SUM(Insulin * Age) - SUM(Insulin) * SUM(Age)) /
        (SQRT((COUNT(*) * SUM(Insulin * Insulin) - SUM(Insulin) * SUM(Insulin))) *
        SQRT((COUNT(*) * SUM(Age * Age) - SUM(Age) * SUM(Age)))) AS Correlation
        FROM diabetes
        WHERE Outcome = 1

'''

re_f9 = pd.read_sql_query(f9, conn)
re_f9

Correlation
0     0.059005

In [ ]:
# Como o índice de massa corporal (IMC) influencia a ocorrência de diabetes em diferentes faixas etárias?

f10 = '''
SELECT
    CASE
        WHEN Age BETWEEN 20 AND 30 THEN '20-30'
        WHEN Age BETWEEN 31 AND 40 THEN '31-40'
        WHEN Age BETWEEN 41 AND 50 THEN '41-50'
        WHEN Age BETWEEN 51 AND 60 THEN '51-60'
        ELSE 'Mais de 60'
    END AS Faixa_Etária,
    CASE
        WHEN Outcome = 1 THEN 'Diabético'
        ELSE 'Não Diabético'
    END AS Categoria,
    AVG(BMI) AS Média_IMC,
    COUNT(*) AS Total_Pacientes
FROM
    diabetes
GROUP BY
    Faixa_Etária, Categoria
ORDER BY
    Faixa_Etária, Categoria
'''


re_f10 = pd.read_sql_query(f10, conn)
re_f10

Faixa_Etária      Categoria  Média_IMC  Total_Pacientes
0        20-30      Diabético  35.890782              358
1        20-30  Não Diabético  29.289264              978
2        31-40      Diabético  35.757042              284
3        31-40  Não Diabético  31.229412              272
4        41-50      Diabético  35.090099              202
5        41-50  Não Diabético  33.246809              141
6        51-60      Diabético  30.148958               96
7        51-60  Não Diabético  31.222807               57
8   Mais de 60      Diabético  30.545000               20
9   Mais de 60  Não Diabético  28.700000               52

# Conclusões finais a partir das consultas anteriores:

- As idade de 25 e 29 anos foram as idades que mais registraram casos para diabetes.
- As idades 67 e 34 anos foram as idades que menos registraram casos para diabetes.
- Com base nas consultas, nota-se uma tendência clara: quanto maior o número de gestações, maior a prevalência de diabetes. Com 100% das mulheres com mais de 14 gestações apresentando diabetes, essa proporção diminui para 84% com 11 gestações e para 62% com 8 gestações. Portanto, fica evidente que o aumento do número de gestações está associado a um maior risco de desenvolver diabetes.
- Fora constatado que não há uma correlação entre os níveis de glicose no sangue e a pressão arterial de pacientes diabéticos, onde o resultado do coeficiente de Pearson é 0.03.
- Fora observado uma correlação positiva moderada entre a espessura da pele e os indíces de IMC, onde o resultado do coeficiente de Pearson é 0.30. Isso significa que, em geral, à medida que o IMC aumenta, a espessura da pele tende a aumentar também, e vice-versa, mas não de forma muito forte.
- Observa-se um maior nível de insulina no sangue para usuários que possuem obesidade de grau 1, grau 2 e grau 3. Essa conexão entre obesidade e hiperinsulinemia exibe a importância de estratégias de prevenção e manejo da obesidade para informar os riscos associados, como o desenvolvimento de diabetes tipo 2 e outras complicações metabólicas.
- Observou-se que não existe correlação entre a idade dos pacientes e a espessura da pele.
- Nos dados da amostra, usuários diabéticos possuem uma espessura da pele maior comparado a usuários não diabéticos.
- Observou-se que não existe correlação entre a idade dos pacientes e níveis de insulina no sangue.
- Conclui-se que existe uma associação entre diabetes e índice de massa corporal (IMC), independente da faixa etária, com os diabéticos tendendo a ter um IMC maior em comparação com os não diabéticos. Além disso, observa-se uma tendência de redução do IMC à medida que as faixas etárias avançam. Essas informações mostram a importância do controle do peso corporal e da prevenção da obesidade como medidas essenciais na gestão e na prevenção do diabetes, independente da idade.


# Estratégias de prevenção com base nas conclusões da análise:


- Educação e conscientização pública: Implementar programas educacionais sobre os fatores de risco para diabetes, incluindo obesidade, número de gestações, e índice de massa corporal. Esses programas podem incluir informações sobre dieta saudável, exercícios físicos regulares e controle do peso corporal.

- Rastreamento e diagnóstico precoce: Estabelecer protocolos de rastreamento regulares para diabetes, especialmente em faixas etárias com maior incidência, como entre 25 e 29 anos. Isso pode incluir exames de glicose no sangue, medições de pressão arterial e avaliação do IMC. Quanto mais cedo o diagnóstico for feito, mais cedo o tratamento pode começar, o que pode reduzir complicações a longo prazo.

- Avaliação regular da espessura da pele e IMC: Introduzir protocolos para avaliação regular da espessura da pele e do IMC em pacientes diabéticos, como parte de exames de rotina. Isso pode ajudar a identificar indivíduos com maior risco de complicações e direcionar intervenções preventivas específicas.

- Integração de cuidados de saúde: Promover uma abordagem multidisciplinar e integrada no cuidado de pacientes diabéticos, envolvendo médicos de várias especialidades, enfermeiros, nutricionistas, e outros profissionais de saúde. Isso pode melhorar a coordenação do tratamento, a adesão do paciente e os resultados de saúde a longo prazo.
